Mathlib/Algebra/Hom/Group.lean

```diff
@@ -0,0 +1,1692 @@
+  by rw [← map_mul, h, map_one]
+#align map_mul_eq_one map_mul_eq_one
+
+@[to_additive]
+theorem map_div' [DivInvMonoid G] [DivInvMonoid H] [MonoidHomClass F G H] (f : F) (hf : ∀ a, f a⁻¹ = (f a)⁻¹)
+    (a b : G) : f (a / b) = f a / f b := by rw [div_eq_mul_inv, div_eq_mul_inv, map_mul, hf]
+#align map_div' map_div'
+
+/-- Group homomorphisms preserve inverse. -/
+@[simp, to_additive "Additive group homomorphisms preserve negation."]
+theorem map_inv [Group G] [DivisionMonoid H] [MonoidHomClass F G H] (f : F) (a : G) : f a⁻¹ = (f a)⁻¹ :=
+  eq_inv_of_mul_eq_one_left <| map_mul_eq_one f <| inv_mul_self _
+#align map_inv map_inv
+
+/-- Group homomorphisms preserve division. -/
+@[simp, to_additive "Additive group homomorphisms preserve subtraction."]
+theorem map_mul_inv [Group G] [DivisionMonoid H] [MonoidHomClass F G H] (f : F) (a b : G) :
+    f (a * b⁻¹) = f a * (f b)⁻¹ := by rw [map_mul, map_inv]
+#align map_mul_inv map_mul_inv
+
+/-- Group homomorphisms preserve division. -/
+@[simp, to_additive "Additive group homomorphisms preserve subtraction."]
+theorem map_div [Group G] [DivisionMonoid H] [MonoidHomClass F G H] (f : F) : ∀ a b, f (a / b) = f a / f b :=
+  map_div' _ <| map_inv f
+#align map_div map_div
+
+-- Porting note: I had some with `to_additive` here,
+-- so have just given separate proofs.
+@[simp]
+theorem map_pow [Monoid G] [Monoid H] [MonoidHomClass F G H] (f : F) (a : G) : ∀ n : ℕ, f (a ^ n) = f a ^ n
+  | 0 => by rw [pow_zero, pow_zero, map_one]
+  | n + 1 => by rw [pow_succ, pow_succ, map_mul, map_pow f a n]
+#align map_pow map_pow
+
+@[simp]
+theorem map_nsmul [AddMonoid G] [AddMonoid H] [AddMonoidHomClass F G H] (f : F) : ∀ (n : ℕ) (a : G), f (n • a) = n • f a
+  | 0, a => by rw [zero_nsmul, zero_nsmul, map_zero]
+  | n + 1, a => by rw [succ_nsmul, succ_nsmul, map_add, map_nsmul f n a]
+#align map_nsmul map_nsmul
+
+attribute [to_additive_reorder 8] map_pow
+
+@[to_additive]
+theorem map_zpow' [DivInvMonoid G] [DivInvMonoid H] [MonoidHomClass F G H] (f : F) (hf : ∀ x : G, f x⁻¹ = (f x)⁻¹)
+    (a : G) : ∀ n : ℤ, f (a ^ n) = f a ^ n
+  | (n : ℕ) => by rw [zpow_coe_nat, map_pow, zpow_coe_nat]
+  | Int.negSucc n => by rw [zpow_neg_succ_of_nat, hf, map_pow, ← zpow_neg_succ_of_nat]
+#align map_zpow' map_zpow'
+
+-- to_additive puts the arguments in the wrong order, so generate an auxiliary lemma, then
+-- swap its arguments.
+/-- Group homomorphisms preserve integer power. -/
+@[simp]
+theorem map_zpow [Group G] [DivisionMonoid H] [MonoidHomClass F G H] (f : F) (g : G) (n : ℤ) : f (g ^ n) = f g ^ n :=
+  map_zpow' f (map_inv f) g n
+#align map_zpow map_zpow
+
+/-- Additive group homomorphisms preserve integer scaling. -/
+theorem map_zsmul [AddGroup G] [SubtractionMonoid H] [AddMonoidHomClass F G H] (f : F) (n : ℤ) (g : G) :
+    f (n • g) = n • f g :=
+  MapZsmul.aux f g n
+#align map_zsmul map_zsmul
+
+attribute [to_additive_reorder 8, to_additive] map_zpow
+
+end mul_one
+
```

Comment:
Is this all commented out because you hadn't got to it? Better to leave it uncommented, please. You can leave a note if you need a "bookmark".

Reply:
Yes I comment the lines out to avoid constant compiling, but I can uncomment them before commits.